components/merkledb/src/proto/mod.rs

```diff
@@ -0,0 +1,201 @@
+
+        for entry in entries {
+            map_proof = match entry.1 {
+                Some(value) => map_proof.add_entry(entry.0, value),
+                None => map_proof.add_missing(entry.0),
+            };
+        }
+
+        Ok(map_proof)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use exonum_crypto::{proto::types, PublicKey};
+    use exonum_proto::ProtobufConvert;
+    use std::fmt;
+
+    use crate::{
+        proto, BinaryKey, BinaryValue, Database, MapProof, ObjectHash, ProofMapIndex, TemporaryDB,
+    };
+    use protobuf::RepeatedField;
+
+    #[test]
+    fn serialize_map_proof() {
+        let db = TemporaryDB::default();
+        let storage = db.fork();
+
+        let mut table = ProofMapIndex::new("index", &storage);
+
+        let proof = table.get_proof(0);
+        assert_proof_roundtrip(proof);
+
+        for i in 0..10 {
+            table.put(&i, i);
+        }
+
+        let proof = table.get_proof(5);
+        assert_proof_roundtrip(proof);
+
+        let proof = table.get_multiproof(5..15);
+        assert_proof_roundtrip(proof);
+    }
+
+    fn assert_proof_roundtrip<K, V>(proof: MapProof<K, V>)
+    where
+        K: BinaryKey + ObjectHash + fmt::Debug,
+        V: BinaryValue + ObjectHash + fmt::Debug,
+        MapProof<K, V>: ProtobufConvert + PartialEq,
+    {
+        let pb = proof.to_pb();
+        let deserialized: MapProof<K, V> = MapProof::from_pb(pb).unwrap();
+        let checked_proof = deserialized
+            .check()
+            .expect("deserialized proof is not valid");
+
+        assert_eq!(proof, deserialized);
+        assert_eq!(
+            checked_proof.index_hash(),
+            proof.check().unwrap().index_hash()
+        );
+    }
+
+    #[test]
+    #[should_panic]
+    fn map_proof_malformed_serialize() {
+        let mut proof = proto::MapProof::new();
+        let mut proof_entry = proto::MapProofEntry::new();
+        proof_entry.set_proof_path(vec![0_u8; 33]);
+        proof.set_proof(RepeatedField::from_vec(vec![proof_entry]));
+
+        let res = MapProof::<u8, u8>::from_pb(proof.clone());
+        assert!(res
+            .unwrap_err()
+            .to_string()
+            .contains("Not valid proof path"));
+
+        let mut proof_entry = proto::MapProofEntry::new();
+        let mut hash = types::Hash::new();
+        hash.set_data(vec![0_u8; 31]);
+        proof_entry.set_hash(hash);
+        proof_entry.set_proof_path(vec![0_u8; 34]);
+        proof.set_proof(RepeatedField::from_vec(vec![proof_entry]));
+
+        let res = MapProof::<u8, u8>::from_pb(proof.clone());
+        assert!(res.unwrap_err().to_string().contains("Wrong Hash size"));
+
+        let mut entry = proto::OptionalEntry::new();
+        entry.set_key(vec![0_u8; 31]);
+        proof.clear_proof();
+        proof.set_entries(RepeatedField::from_vec(vec![entry]));
+
+        // TODO: will panic at runtime, should change BinaryKey::read signature
```

Comment:
TODO without ECR (and maybe explain a little bit more, it doesn't seem clear to me?).